solang-parser/src/pt.rs

```diff
@@ -1,1003 +1,1688 @@
+    }
+
+    /// Returns mutable references to the components of this expression.
+    ///
+    /// See also [`Expression::components`].
+    #[inline]
+    pub fn components_mut(&mut self) -> (Option<&mut Self>, Option<&mut Self>) {
+        use Expression::*;
+        expr_components!(self)
+    }
+
+    /// Returns whether this expression can be split across multiple lines.
+    #[inline]
+    pub const fn is_unsplittable(&self) -> bool {
+        use Expression::*;
+        matches!(
+            self,
+            BoolLiteral(..)
+                | NumberLiteral(..)
+                | RationalNumberLiteral(..)
+                | HexNumberLiteral(..)
+                | StringLiteral(..)
+                | HexLiteral(..)
+                | AddressLiteral(..)
+                | Variable(..)
+                | This(..)
+        )
+    }
+
+    /// Returns whether this expression has spaces around it.
+    #[inline]
+    pub const fn has_space_around(&self) -> bool {
+        use Expression::*;
+        !matches!(
+            self,
+            PostIncrement(..)
+                | PreIncrement(..)
+                | PostDecrement(..)
+                | PreDecrement(..)
+                | Not(..)
+                | Complement(..)
+                | UnaryPlus(..)
+                | Negate(..)
+        )
+    }
 }
 
+/// A parameter.
+///
+/// `<ty> [storage] <name>`
 #[derive(Debug, PartialEq, Eq, Clone)]
 #[cfg_attr(feature = "pt-serde", derive(Serialize, Deserialize))]
 pub struct Parameter {
+    /// The code location.
     pub loc: Loc,
+    /// The type.
     pub ty: Expression,
+    /// The optional memory location.
     pub storage: Option<StorageLocation>,
+    /// The optional identifier.
     pub name: Option<Identifier>,
 }
 
+/// Function mutability.
 #[derive(Debug, PartialEq, Eq, Clone)]
 #[cfg_attr(feature = "pt-serde", derive(Serialize, Deserialize))]
 pub enum Mutability {
+    /// `pure`
     Pure(Loc),
+
+    /// `view`
     View(Loc),
-    Constant(Loc),
-    Payable(Loc),
-}
 
-impl fmt::Display for Mutability {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self {
-            Mutability::Pure(_) => write!(f, "pure"),
-            Mutability::Constant(_) | Mutability::View(_) => write!(f, "view"),
-            Mutability::Payable(_) => write!(f, "payable"),
-        }
-    }
-}
+    /// `constant`
+    Constant(Loc),
 
-impl CodeLocation for Mutability {
-    fn loc(&self) -> Loc {
-        match self {
-            Mutability::Pure(loc)
-            | Mutability::Constant(loc)
-            | Mutability::View(loc)
-            | Mutability::Payable(loc) => *loc,
-        }
-    }
+    /// `payable`
+    Payable(Loc),
 }
 
+/// Function visibility.
+///
+/// Deprecated for [FunctionTy] other than `Function`.
 #[derive(Debug, PartialEq, Eq, Clone)]
 #[cfg_attr(feature = "pt-serde", derive(Serialize, Deserialize))]
+#[repr(u8)] // for cmp; order of variants is important
```

Comment:
Why do you need the representation to be `u8`?

Reply:
it's needed to add the Ord implementations by using only the discriminant, see the cmp file. But I realize it might be better off just deriving the implementations